jsprit-core/src/main/java/com/graphhopper/jsprit/core/algorithm/recreate/DefaultScorer.java

```diff
@@ -42,53 +41,75 @@ public DefaultScorer(VehicleRoutingProblem vrp) {
     }
 
     public void setTimeWindowParam(double tw_param) {
-        this.timeWindowParam = tw_param;
+        timeWindowParam = tw_param;
     }
 
     public void setDepotDistanceParam(double depotDistance_param) {
-        this.depotDistanceParam = depotDistance_param;
+        depotDistanceParam = depotDistance_param;
     }
 
     @Override
     public double score(InsertionData best, Job job) {
-        double score;
-        if (job instanceof Service) {
-            score = scoreService(best, job);
-        } else if (job instanceof Shipment) {
-            score = scoreShipment(best, job);
-        } else throw new IllegalStateException("not supported");
-        return score;
-    }
-
-    private double scoreShipment(InsertionData best, Job job) {
-        Shipment shipment = (Shipment) job;
-        double maxDepotDistance_1 = Math.max(
-            getDistance(best.getSelectedVehicle().getStartLocation(), shipment.getPickupLocation()),
-            getDistance(best.getSelectedVehicle().getStartLocation(), shipment.getDeliveryLocation())
-        );
-        double maxDepotDistance_2 = Math.max(
-            getDistance(best.getSelectedVehicle().getEndLocation(), shipment.getPickupLocation()),
-            getDistance(best.getSelectedVehicle().getEndLocation(), shipment.getDeliveryLocation())
-        );
-        double maxDepotDistance = Math.max(maxDepotDistance_1, maxDepotDistance_2);
-        double minTimeToOperate = Math.min(shipment.getPickupTimeWindow().getEnd() - shipment.getPickupTimeWindow().getStart(),
-            shipment.getDeliveryTimeWindow().getEnd() - shipment.getDeliveryTimeWindow().getStart());
+        Vehicle selectedVehicle = best.getSelectedVehicle();
+        double maxFromStart = job.getAllLocations().stream()
+                .mapToDouble(l -> getDistance(selectedVehicle.getStartLocation(), l))
+                .max()
+                .orElse(0d);
+        double maxToEnd = job.getAllLocations().stream()
+                .mapToDouble(l -> getDistance(selectedVehicle.getEndLocation(), l))
+                .max()
+                .orElse(0d);
+        double maxDepotDistance = Math.max(maxFromStart, maxToEnd);
+        double minTimeToOperate = job.getOperationTimeWindows().stream()
+                .mapToDouble(tw -> tw.getEnd() - tw.getStart())
+                .min()
+                .orElse(0d);
```

Comment:
I've messured stream API in our application and was surprised its performance. Although I accept that the most mission critical code should be revised if needed.
Also we have to accept, that a unified solution always has an overhead, but gives flexibility. 
I propose to follow what Knuth said: "Do not use premature optimalization." 
Let's make the code run well, let's make it flexible, then revise the possible bottlenecks.